src/cls/async-hooks.ts

```diff
@@ -64,51 +98,80 @@ export class AsyncHooksCLS<Context extends {}> implements CLS<Context> {
   }
 
   enable(): void {
-    this.currentContext.value = this.defaultContext;
+    this.contexts = {};
     this.hook.enable();
     this.enabled = true;
   }
 
   disable(): void {
-    this.currentContext.value = this.defaultContext;
+    this.contexts = {};
     this.hook.disable();
     this.enabled = false;
   }
 
   getContext(): Context {
-    return this.currentContext.value;
+    // We don't store this.defaultContext directly in this.contexts.
+    // Getting undefined when looking up this.contexts means that it wasn't
+    // set, so return the default context.
+    const current = this.contexts[this.ah.executionAsyncId()];
+    return current ? current.value : this.defaultContext;
   }
 
   setContext(value: Context): void {
```

Comment:
Comment for a follow-on PR: it would be good to merge this into `runWithNewContext` and implement a single `runInContext` function.

Reply:
Added a TODO.